modules/rosetta/src/main/scala/org/tessellation/rosetta/server/model/dag/GenerateCodecs.scala

```diff
@@ -0,0 +1,20 @@
+package org.tessellation.rosetta.server.model.dag
+
+import scala.reflect.io.Path
+
+object GenerateCodecs {
+
+  def run(): Unit =
+    scala.tools.nsc.io.File
```

Comment:
We are using `File` from fs2. I guess you could use it here as well.

Reply:
Can you please point me at a tutorial on using FS2 with a directory? So far, I've figured out

```scala
Files[F].walk(Path("modules/rosetta/src/main/scala/org/tessellation/rosetta/server/model"))
```

However, I don't know how to check if a `Path` is a file and I can't call `foreach`.

Reply:
> Switching over to FS2 doesn't really provide real value.

It does provide a value. Using the same approach across the whole codebase - fs2 and cats-effect for all IO operations makes the code consistent and interoperable.

Reply:
I will need assistance with this. We can go over it when we meet @tbekas next week. Alternatively, please point me to a good tutorial.

Reply:
@cngo-github both [cats-effect ](https://typelevel.org/cats-effect/)and [fs2](https://fs2.io/#/) have great documentation with plenty of examples and use-cases.

Reply:
This is what I was able to figure out:

```
def run[F[_]: Files: Applicative](): Unit = {
    val generatedCodeDirectory = Path("modules/rosetta/src/main/scala/org/tessellation/rosetta/api/generated")
    Files[F].list(generatedCodeDirectory).filter(path => path.fileName.toString.endsWith(".scala")).foreach { entry =>
      val fileName = entry.fileName.toString.replaceFirst(".scala", "")
      val genString = s"""implicit lazy val ${fileName}Encoder = deriveConfiguredMagnoliaEncoder[${fileName}]
                         |implicit lazy val ${fileName}Decoder = deriveConfiguredMagnoliaDecoder[${fileName}]""".stripMargin
      println(genString)
      Applicative[F].pure(())
    }
  }
```

Reply:
@cngo-github using functional programming, we use effects to encapsulate the side-effects and delay their execution. for example: `println(genString)` is a side effect that will be executed immediately, but wrapping it into `IO(println(genString))` will create a declaration of run, so you can freely pass it further and execute on demand.

the whole codebase is running as a single "run" command, by chaining all the small sub-programs altogether.

that being said, you should chain (`flatMap`) all the operations.

`Applicative[F].pure(())` makes no sense because it does nothing.
`foreach` is imperative and it expects a unit type, and it expects a side-effect. instead of foreach loop, you should use fold or reduce.

anyway, it's not worth trying (other than learning purpose) because this code should be changed anyway to generate classes using template

Reply:
After discussing this with @ryle-goehausen , I'll be removing this file. I will work with @tbekas on Friday to automate this if we can do so relatively quickly. If not, we'll just leave this out of the code base.